File: aws/src/handlers/user_handler.py
import json
import boto3
import os
import base64
from cors_utils import cors_response, cors_error_response, handle_preflight_request, extract_origin_from_event
from datetime import datetime
from decimal import Decimal
from botocore.exceptions import ClientError

# Initialize AWS clients
dynamodb = boto3.resource('dynamodb')
s3 = boto3.client('s3')

# Environment variables
USERS_TABLE = os.environ['DYNAMODB_TABLE_USERS']
BOOKINGS_TABLE = os.environ['DYNAMODB_TABLE_BOOKINGS']
S3_BUCKET = os.environ.get('S3_BUCKET', 'dogotel-images')

users_table = dynamodb.Table(USERS_TABLE)
bookings_table = dynamodb.Table(BOOKINGS_TABLE)

def lambda_handler(event, context):
    """Main handler for user-related endpoints"""
    
    # Handle preflight requests
    if event['httpMethod'] == 'OPTIONS':
        return handle_preflight_request(event)
    
    try:
        origin = extract_origin_from_event(event)
        
        path = event['pathParameters']['proxy'] if event.get('pathParameters') and event['pathParameters'].get('proxy') else ''
        http_method = event['httpMethod']
        
        print(f"User handler - Path: {path}, Method: {http_method}")
        
        if path.startswith('user/') and http_method == 'GET':
            email = path.split('/')[-1]
            return handle_get_user_profile(email, event)
        elif path.startswith('bookings/') and http_method == 'GET':
            email = path.split('/')[-1]
            return handle_get_booking_history(email, event)
        else:
            return cors_error_response(404, 'Endpoint not found', origin)
            
    except Exception as e:
        print(f"User handler error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_get_user_profile(email, event):
    """Handle GET /api/user/:email - exact format from Ran_dev.txt"""
    try:
        origin = extract_origin_from_event(event)
        
        # Get user from database
        response = users_table.get_item(Key={'email': email})
        
        if 'Item' not in response:
            return cors_response(200, {
                'success': False,
                'error': 'User not found'
            }, origin)
        
        user = response['Item']
        
        return cors_response(200, {
            'success': True,
            'user': {
                'username': user.get('name', user.get('username', '')),
                'birthdate': user.get('birthdate', ''),
                'email': user.get('email'),
                'photo': user.get('photo', 'https://via.placeholder.com/150')
            }
        }, origin)
        
    except Exception as e:
        print(f"Get user profile error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_get_booking_history(email, event):
    """Handle GET /api/bookings/:email - exact format from Ran_dev.txt"""
    try:
        origin = extract_origin_from_event(event)
        
        # Get user's bookings from database
        response = bookings_table.query(
            IndexName='UserBookingsIndex',
            KeyConditionExpression='userEmail = :userEmail',
            ExpressionAttributeValues={
                ':userEmail': email
            }
        )
        
        if not response['Items']:
            return cors_response(200, {
                'success': False,
                'error': 'No bookings found for this user'
            }, origin)
        
        # Transform bookings to match Ran_dev.txt format
        history = []
        for booking in response['Items']:
            room_data = booking.get('room', {})
            dining_data = booking.get('dining', {})
            services_data = booking.get('services', [])
            
            dogs = room_data.get('dogs', [])
            dog_names = [dog.get('name', '') for dog in dogs]
            
            # Calculate date range
            start_date = room_data.get('startDate', '')
            end_date = room_data.get('endDate', '')
            dates = f"{start_date} to {end_date}" if start_date and end_date else 'Unknown dates'
            
            # Get dining selection
            dining_selection = dining_data.get('title', 'None') if dining_data else 'None'
            
            # Get services selection
            services_selection = ', '.join([s.get('title', '') for s in services_data]) if services_data else 'None'
            
            history.append({
                'bookingNumber': booking.get('bookingId', ''),
                'dates': dates,
                'dogs': dog_names,
                'diningSelection': dining_selection,
                'servicesSelection': services_selection,
                'totalPrice': booking.get('totalPrice', 0)
            })
        
        return cors_response(200, {
            'success': True,
            'history': history
        }, origin)
        
    except Exception as e:
        print(f"Get booking history error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def upload_image_to_s3(base64_image, key_prefix):
    """Upload base64 image to S3 and return the URL"""
    try:
        # Extract image data from base64 string
        header, encoded = base64_image.split(',', 1)
        image_data = base64.b64decode(encoded)
        
        # Determine file extension from header
        if 'jpeg' in header or 'jpg' in header:
            ext = 'jpg'
        elif 'png' in header:
            ext = 'png'
        else:
            ext = 'jpg'  # default
        
        # Create S3 key
        s3_key = f"{key_prefix}.{ext}"
        
        # Upload to S3
        s3.put_object(
            Bucket=S3_BUCKET,
            Key=s3_key,
            Body=image_data,
            ContentType=f"image/{ext}",
            ACL='public-read'
        )
        
        # Return S3 URL
        return f"https://{S3_BUCKET}.s3.amazonaws.com/{s3_key}"
        
    except Exception as e:
        print(f"Error uploading image to S3: {str(e)}")
        return base64_image  # Return original if upload fails 

File: aws/src/handlers/user_handler.js
const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const {
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
  UpdateCommand,
} = require("@aws-sdk/lib-dynamodb");
const {
  CognitoIdentityProviderClient,
  AdminGetUserCommand,
} = require("@aws-sdk/client-cognito-identity-provider");
const {
  corsResponse,
  corsErrorResponse,
  handlePreflightRequest,
  extractOriginFromEvent,
} = require("./cors_utils");

// Initialize AWS clients
const ddbClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(ddbClient);
const cognitoClient = new CognitoIdentityProviderClient({
  region: process.env.AWS_REGION,
});

// Environment variables
const USERS_TABLE = process.env.USERS_TABLE || "DogotelUsers";
const USER_POOL_ID = process.env.USER_POOL_ID;

exports.handler = async (event, context) => {
  console.log("Event:", JSON.stringify(event, null, 2));

  try {
    const httpMethod = event.httpMethod;
    const path = event.path;
    const origin = extractOriginFromEvent(event);

    console.log(`Processing ${httpMethod} ${path} from origin: ${origin}`);

    // Handle preflight OPTIONS requests
    if (httpMethod === "OPTIONS") {
      return handlePreflightRequest(event, origin);
    }

    if (path.startsWith("/user/") && httpMethod === "GET") {
      // Handle GET /user/:email for user profile
      const email = path.split("/user/")[1];
      return await handleGetUserProfile(event, email);
    } else if (path === "/user/profile-photo" && httpMethod === "POST") {
      // Handle POST /user/profile-photo for updating profile photo
      return await handleUpdateProfilePhoto(event);
    } else if (path === "/user/profile" && httpMethod === "PUT") {
      // Handle PUT /user/profile for updating user profile
      return await handleUpdateUserProfile(event);
    } else {
      return corsErrorResponse(404, "Endpoint not found", origin);
    }
  } catch (error) {
    console.error("Error in user handler:", error);
    return corsErrorResponse(
      500,
      "Internal server error",
      extractOriginFromEvent(event)
    );
  }
};

async function handleGetUserProfile(event, userEmail) {
  try {
    const origin = extractOriginFromEvent(event);
    const decodedEmail = decodeURIComponent(userEmail);

    if (!decodedEmail) {
      return corsErrorResponse(400, "User email is required", origin);
    }

    // Get user from Cognito to ensure we have the latest attributes
    let cognitoUser;
    try {
      cognitoUser = await cognitoClient.send(
        new AdminGetUserCommand({
          UserPoolId: USER_POOL_ID,
          Username: decodedEmail,
        })
      );
    } catch (cognitoError) {
      console.warn(
        "Could not find user in Cognito, will check DynamoDB.",
        cognitoError
      );
    }

    // Get user from DynamoDB
    const dynamoResult = await docClient.send(
      new GetCommand({
        TableName: USERS_TABLE,
        Key: { email: decodedEmail },
      })
    );

    if (!dynamoResult.Item && !cognitoUser) {
      return corsErrorResponse(
        404,
        "User not found in Cognito or DynamoDB",
        origin
      );
    }

    // Merge data from Cognito and DynamoDB
    const cognitoAttributes = {};
    if (cognitoUser && cognitoUser.UserAttributes) {
      cognitoUser.UserAttributes.forEach((attr) => {
        cognitoAttributes[attr.Name] = attr.Value;
      });
    }

    const dynamoData = dynamoResult.Item || {};

    const mergedUser = {
      email: decodedEmail,
      username:
        cognitoAttributes.preferred_username ||
        dynamoData.username ||
        `${cognitoAttributes.given_name} ${cognitoAttributes.family_name}`.trim() ||
        decodedEmail.split("@")[0],
      firstName: cognitoAttributes.given_name || dynamoData.firstName || "",
      lastName: cognitoAttributes.family_name || dynamoData.lastName || "",
      birthdate: cognitoAttributes.birthdate || dynamoData.birthdate || "",
      photo: dynamoData.photo || dynamoData.profilePhoto || "",
      createdAt: cognitoUser
        ? new Date(cognitoUser.UserCreateDate).toISOString()
        : dynamoData.createdAt,
      updatedAt: cognitoUser
        ? new Date(cognitoUser.UserLastModifiedDate).toISOString()
        : dynamoData.updatedAt,
    };

    // If user was in Cognito but not Dynamo, create a record for them
    if (cognitoUser && !dynamoResult.Item) {
      await docClient.send(
        new PutCommand({
          TableName: USERS_TABLE,
          Item: {
            email: mergedUser.email,
            firstName: mergedUser.firstName,
            lastName: mergedUser.lastName,
            username: mergedUser.username,
            birthdate: mergedUser.birthdate,
            profilePhoto: mergedUser.photo,
            createdAt: mergedUser.createdAt,
            updatedAt: mergedUser.updatedAt,
          },
        })
      );
    }

    return corsResponse(
      200,
      {
        success: true,
        user: mergedUser,
      },
      origin
    );
  } catch (error) {
    console.error("Get user profile error:", error);
    return corsErrorResponse(
      500,
      "Internal server error",
      extractOriginFromEvent(event)
    );
  }
}

async function handleUpdateProfilePhoto(event) {
  try {
    const origin = extractOriginFromEvent(event);
    const body = JSON.parse(event.body);
    const { email, photo } = body;

    if (!email || !photo) {
      return corsErrorResponse(400, "Email and photo are required", origin);
    }

    // Update user profile photo in DynamoDB using UpdateCommand to preserve other data
    const result = await docClient.send(
      new UpdateCommand({
        TableName: USERS_TABLE,
        Key: { email: email },
        UpdateExpression: "SET photo = :photo, updated_at = :updated_at",
        ExpressionAttributeValues: {
          ":photo": photo,
          ":updated_at": new Date().toISOString(),
        },
      })
    );

    return corsResponse(
      200,
      {
        success: true,
        message: "Profile photo updated successfully",
        photoUrl: photo,
      },
      origin
    );
  } catch (error) {
    console.error("Update profile photo error:", error);
    return corsErrorResponse(
      500,
      "Internal server error",
      extractOriginFromEvent(event)
    );
  }
}

async function handleUpdateUserProfile(event) {
  try {
    const origin = extractOriginFromEvent(event);
    const body = JSON.parse(event.body);
    const { email, username, birthdate, photo } = body;

    if (!email) {
      return corsErrorResponse(400, "Email is required", origin);
    }

    // Build update expression dynamically
    const updateExpressions = [];
    const expressionAttributeValues = {
      ":updated_at": new Date().toISOString(),
    };

    if (username) {
      updateExpressions.push("username = :username");
      expressionAttributeValues[":username"] = username;
    }

    if (birthdate) {
      updateExpressions.push("birthdate = :birthdate");
      expressionAttributeValues[":birthdate"] = birthdate;
    }

    if (photo !== undefined) {
      updateExpressions.push("photo = :photo");
      expressionAttributeValues[":photo"] = photo;
    }

    updateExpressions.push("updated_at = :updated_at");

    // Update user profile in DynamoDB
    const result = await docClient.send(
      new UpdateCommand({
        TableName: USERS_TABLE,
        Key: { email: email },
        UpdateExpression: `SET ${updateExpressions.join(", ")}`,
        ExpressionAttributeValues: expressionAttributeValues,
        ReturnValues: "ALL_NEW",
      })
    );

    return corsResponse(
      200,
      {
        success: true,
        message: "Profile updated successfully",
        user: {
          username: result.Attributes.username,
          birthdate: result.Attributes.birthdate,
          email: result.Attributes.email,
          photo: result.Attributes.photo || "",
        },
      },
      origin
    );
  } catch (error) {
    console.error("Update user profile error:", error);
    return corsErrorResponse(
      500,
      "Internal server error",
      extractOriginFromEvent(event)
    );
  }
}
